From 81de29d842ccb776c0f77aa3e2b11b07fff0c0e2 Mon Sep 17 00:00:00 2001
From: Allen Pais <allen.pais@oracle.com>
Date: Wed, 18 Sep 2019 22:00:31 +0530
Subject: [PATCH] drm/amdkfd: fix a potential NULL pointer dereference (v2)

alloc_workqueue is not checked for errors and as a result,
a potential NULL dereference could occur.

v2 (Felix Kuehling):
* Fix compile error (kfifo_free instead of fifo_free)
* Return proper error code

# adapted for the amdgpu-dkms driver
--- a/amd/amdkfd/kfd_interrupt.c.orig	2019-07-30 14:24:04.000000000 -0700
+++ b/amd/amdkfd/kfd_interrupt.c	2020-01-07 01:23:26.567040778 -0800
@@ -65,6 +65,11 @@ int kfd_interrupt_init(struct kfd_dev *k
 #else
 	kfd->ih_wq = alloc_workqueue("KFD IH", WQ_HIGHPRI, 1);
 #endif
+	if (unlikely(!kfd->ih_wq)) {
+		kfifo_free(&kfd->ih_fifo);
+		dev_err(kfd_chardev(), "Failed to allocate KFD IH workqueue\n");
+		return -ENOMEM;
+	}
 	spin_lock_init(&kfd->interrupt_lock);
